packages/expo-av/android/src/main/java/expo/modules/av/AVManager.java

```diff
@@ -669,6 +676,135 @@ public void prepareAudioRecorder(final ReadableArguments options, final Promise
+    int type = deviceInfo.getType();
+    String typeStr = String.valueOf(type);
+    if (type == AudioDeviceInfo.TYPE_BUILTIN_MIC) {
+      typeStr = "MicrophoneBuiltIn";
+    } else if (type == AudioDeviceInfo.TYPE_BLUETOOTH_SCO) {
+      typeStr = "BluetoothSCO";
+    } else if (type == AudioDeviceInfo.TYPE_BLUETOOTH_A2DP) {
+      typeStr = "BluetoothA2DP";
+    } else if (type == AudioDeviceInfo.TYPE_TELEPHONY) {
+      typeStr = "Telephony";
+    } else if (type == AudioDeviceInfo.TYPE_WIRED_HEADSET) {
+      typeStr = "MicrophoneWired";
+    }
+    map.putString(RECORDING_INPUT_NAME_KEY, deviceInfo.getProductName().toString());
+    map.putString(RECORDING_INPUT_TYPE_KEY, typeStr);
+    map.putString(RECORDING_INPUT_UID_KEY, String.valueOf(deviceInfo.getId()));
+    return map;
+  }
+
+  @Override
+  public void getCurrentInput(final Promise promise) {
+    AudioDeviceInfo deviceInfo = null;
+    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
+
+      try {
+        // getRoutedDevice() is the most reliable way to return the actual mic input, however it
+        // only returns a valid device when actively recording, and may throw otherwise.
+        // https://developer.android.com/reference/android/media/MediaRecorder#getRoutedDevice()
+        deviceInfo = mAudioRecorder.getRoutedDevice();
+      } catch (Exception e) {
+        // Noop if this throws, try alternate method of determining current input below.
+      }
+
+      // If no routed device is found try preferred device
+      deviceInfo = mAudioRecorder.getPreferredDevice();
+
+      if (deviceInfo == null) {
+        // If no preferred device is found, set it to the first built-in input we can find
+        AudioDeviceInfo[] audioDevices = mAudioManager.getDevices(AudioManager.GET_DEVICES_INPUTS);
+        for (int i = 0; i < audioDevices.length; i++) {
+          AudioDeviceInfo availableDeviceInfo = audioDevices[i];
+          int type = availableDeviceInfo.getType();
+          if (type == AudioDeviceInfo.TYPE_BUILTIN_MIC) {
+            deviceInfo = availableDeviceInfo;
+            mAudioRecorder.setPreferredDevice(deviceInfo);
+            break;
+          }
+        }
+      }
+    } else {
+      promise.reject("E_AUDIO_VERSIONINCOMPATIBLE", "Getting current input is unsupported on" +
+        " Android devices running SDK < 24.");
+      return;
+    }
+    if (deviceInfo != null) {
+      final Bundle map = getMapFromDeviceInfo(deviceInfo);
+      promise.resolve(map);
+    } else {
+      promise.reject(null);
+    }
+  }
+
+  @Override
+  public void getAvailableInputs(final Promise promise) {
+    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+      ArrayList<Bundle> devices = new ArrayList();
+      AudioDeviceInfo[] audioDevices = mAudioManager.getDevices(AudioManager.GET_DEVICES_INPUTS);
+      for (int i = 0; i < audioDevices.length; i++) {
+        AudioDeviceInfo deviceInfo = audioDevices[i];
```

Comment:
A more concise syntax:

```suggestion
      for (AudioDeviceInfo deviceInfo : audioDevices) {
```